modules/packages/slackpkg

```diff
@@ -0,0 +1,188 @@
+    OPTIONS="$*"
+
+    for OPT in ${OPTIONS} ; do
+        eval `parse_line ${OPT}`
+        case "${KEYWORD}" in
+            File*)
+                if [ "${CMD}" = 'file-install' ] ; then
+                    PKG="${PKG} ${VALUE}"
+                else
+                    fatal "Bad Input: '${OPT}' (should be Name=...)."
+                fi
+                VER=''
+                ;;
+            Name*)
+                if [ "${CMD}" != 'file-install' ] ; then
+                    PKG="${PKG} ${VALUE}"
+                else
+                    fatal "Bad Input: '${OPT}' (should be File=...)"
+                fi
+                VER=''
+                ;;
+            Version*)
+                PKG="${PKG}-${VALUE}"
+                ;;
+            Architecture*)
+                if [ -n "${VALUE}" ] &&
+                       [ "${VALUE}" != "${DEFARCH}" ] ; then
+                    fatal "Bad Input: Architecture ${VALUE} does not match ${DEFARCH}."
+                fi
+                ;;
+            PackageType*)
+                TYP="${VALUE}"
+                ;;
+            *)
+                fatal "Invalid input: '${INPUT}'."
+                ;;
+        esac
+    done
+}
+
+get_package_data() {
+    if [ -n "${NAM}" ] && [ -z "${PKG}" ] ; then
+        PKG="${NAM}"
+    fi
+    if [ -n "${PKG}" ] ; then
+        case "${PKG}" in
+            /*|./*) # File Path.  It's a file
+                TYP='file'
+                if [ ! -e "${PKG}" ] ; then
+                    fatal "No such file: '${PKG}'."
+                fi
+                ;;
+            http://*|ftp://|file://) # URL, it's a file.
+                TYP='file'
+                ;;
+            *)
+                TYP='repo'
+                ;;
+        esac
+        echo "PackageType=${TYP}"
+        if [ "${TYP}" = 'repo' ] ; then
+            if [ -z "`echo ${PKGVER} | grep -E '^[0-9]'`" ] ; then
+                if [ "${PKGNAM}" != "${PKGVER}" ] ;  then
+                    PKGNAM="${PKGNAM}-${PKGVER}"
+                fi
+                PKGVER=''
+            fi
+            # Returning version for PackageType=repo is not advised
+            # https://docs.cfengine.com/latest/reference-standard-library-package_modules.html
+            # if [ -n "${PKGVER}" ] ; then
+            #     echo "Version=${PKGVER}"
+            # fi
+            echo "Name="${PKG}
+        else
+          echo "Name="$(echo ${PKG} |sed -e 's/-[0-9].*$//' -e 's/[^/]+\(.*$\)/\1/' -e 's/.*\///')
+          echo ${PKG} |sed -e 's/.*\///' -e 's/.*-\([_a-z0-9.?]\+\)-\(.*\)-.*/Version=\1\'$'\nArchitecture=\\2/'
+        fi
+    fi
+}
+
+if [ "${CMD}" = 'supports-api-version' ] ; then
+    echo '1'
+    exit 0
+else
+    read_options
+fi
+
+case "${CMD}" in
+    get-package-data)
+        get_package_data
+        ;;
+    list-installed)
+        ls -l /var/log/packages/ |sed -e 's/.* \([^ ]\+\)-\([_a-w0-9.?]\+\)-\(.*\)-.*/Name=\1\'$'\nVersion=\\2\\nArchitecture=\\3/'
+        ;;
```

Comment:
I tried the basic nmap absent example you give in the docs PR https://github.com/cfengine/documentation/pull/1823 and get some errors.

```
root@slackami:/var/cfengine/masterfiles# vim services/main.cf
root@slackami:/var/cfengine/masterfiles# cf-promises services/main.cf
root@slackami:/var/cfengine/masterfiles# cf-agent -Klf update.cf
root@slackami:/var/cfengine/masterfiles# cf-agent -KI
   error: Unsupported response from package module: total 13276
   error: Unsupported response from package module: -rw-r--r-- 1 root root   10515
Feb 22  2014 bind-9.9.3_P2-x86_64-1
   error: Unsupported response from package module: -rw-r--r-- 1 root root    1443
Feb 22  2014 cdparanoia-III_10.2-x86_64-1
   error: Unsupported response from package module: -rw-r--r-- 1 root root    1879
Feb 22  2014 dhcp-4.2.5_P1-x86_64-1
    info: Error updating packages cache.
    info: Some error occurred while updating available updates cache.
   error: Can not update cache.
   error: Unsupported response from package module: total 13232
   error: Unsupported response from package module: -rw-r--r-- 1 root root   10515
Feb 22  2014 bind-9.9.3_P2-x86_64-1
   error: Unsupported response from package module: -rw-r--r-- 1 root root    1443
Feb 22  2014 cdparanoia-III_10.2-x86_64-1
   error: Unsupported response from package module: -rw-r--r-- 1 root root    1879
Feb 22  2014 dhcp-4.2.5_P1-x86_64-1
    info: Error updating packages cache.
    info: Can not update installed packages cache after package installation
   error: Error removing package 'nmap'
```

I'm wondering if you could change the version capture to ```([_a-zA-Z0-9.?]+)``` and somehow get rid of the ```total 13232``` output of ls -l maybe with a ```grep -v "^total"``` in between the ls and sed commands? So something like

```ls -l /var/log/packages/ | grep -v "^total" |sed -e 's/.* \([^ ]\+\)-\([_a-zA-Z0-9.?]\+\)-\(.*\)-.*/Name=\1\'$'\nVersion=\\2\\nArchitecture=\\3/'```

That change fixes things up for me pretty well.
+        ;;

Reply:
Fixed this by changing the ls option to -1, this gives less output for sed to filter and simplifies the regex. Also changed the version capture as you recommended. Strangely this issue never occured to me..